bootstrap/bootstrap-kubestellar.sh

```diff
@@ -194,67 +248,131 @@ if [ "$folder" == "" ]; then
+echo "Ensure kcp is installed"
+echo "************************************************"
+if [ "$(kcp_installed)" == "true" ]; then
+    echo "kcp found in the PATH at '$(which kcp)' ... skip installation."
+else
+    echo "kcp not found in the PATH."
+    if [ "$(kcp_running)" == "true" ]; then
+        echo "kcp process is running already: pid=$(pgrep kcp) ... add kcp folder to the PATH or stop kcp ... exiting."
+        exit 2
     fi
-    bash <(curl -s https://raw.githubusercontent.com/kcp-dev/edge-mc/main/bootstrap/install-kcp-with-plugins.sh) --version $kcp_version --os $os_type --arch $arch_type --ensure-folder  $folder/kcp $verbose $flagx
-    if [[ ! ":$PATH:" == *":$(get_full_path $folder/kcp/bin):"* ]]; then
-        export PATH=$PATH:$(get_full_path $folder/kcp/bin)
+    ensure_folder "$folder/kcp"
+    kcp_folder=$(get_full_path "$folder/kcp")
+    kcp_bin_folder="$kcp_folder/bin"
+    echo "Downloading kcp+plugins $kcp_version $os_type/$arch_type..."
+    kcp_download
+    echo "Installing kcp+plugins into '$kcp_folder'..."
+    kcp_install
+    if [[ ! ":$PATH:" == *":$kcp_bin_folder:"* ]]; then
+        export PATH=$kcp_bin_folder:$PATH
+        echo "Add kcp folder to the PATH: export PATH=\"$kcp_bin_folder:\$PATH\""
+        user_exports="$user_exports"$'\n'"export PATH=\"$kcp_bin_folder:\$PATH\""
     fi
 fi
 
 # Ensure kcp is running
-if [ "$(kcp_running)" == "false" ]; then
+echo "************************************************"
+echo "Ensure kcp is running"
+echo "************************************************"
+if [ "$(kcp_running)" == "true" ]; then
+    echo "kcp process is running already: pid=$(pgrep kcp) ... skip running."
+    if [ "$(kubeconfig_valid)" == "false" ]; then
+        echo "KUBECONFIG environment variable is not set correctly: KUBECONFIG='$KUBECONFIG' ... exiting!"
+        exit 3
+    fi
+    echo "Using 'KUBECONFIG=$KUBECONFIG'"
+    echo "Waiting for kcp to be ready... it may take a while"
+    until $(kcp_ready)
+    do
+        sleep 1
+    done
+    if [ "$(kcp_version)" != "$kcp_version" ]; then
+        echo "kcp running version $(kcp_version) does not match the desired version $kcp_version ... exiting!"
+        exit 4
+    else
+        echo "kcp version $(kcp_version) ... ok"
+    fi
+else
     if [ "$kcp_address" == "" ]; then
-        if [ "$verbose" != "" ]; then
-            echo "Starting kcp..."
-        fi
+        echo "Running kcp... logfile=$PWD/kcp_log.txt"
         kcp start >& kcp_log.txt &
     else
-        if [ "$verbose" != "" ]; then
-            echo "Starting kcp bound to address $kcp_address..."
-        fi
+        echo "Running kcp bound to address $kcp_address... logfile=$PWD/kcp_log.txt"
         kcp start --bind-address $kcp_address >& kcp_log.txt &
     fi
-    export KUBECONFIG="$(pwd)/.kcp/admin.kubeconfig"
+    export KUBECONFIG="$PWD/.kcp/admin.kubeconfig"
+    echo "Waiting for kcp to be ready... it may take a while"
```

Comment:
I see two copies of the same wait.  Can this be re-organized so that the different paths converge before the waiting?

Reply:
probably